### ISSUE TITLE
Typescript 2.4 breaks support of enum

### ISSUE THREAD

**Issue by noemi-salaun** [weichx/cerialize#61]
With 2.4, Typescript add some check about weak types.

See [the related changelog](https://github.com/Microsoft/TypeScript/wiki/What%27s-new-in-TypeScript#weak-type-detection).

If we look at the **deserializeAs** annotation signature

```ts
export declare function deserializeAs(keyNameOrType: string | Function | ISerializable, keyName?: string): any;
```

And its usage with enums

```ts

// in my.enum.ts
export enum MyEnum { foo, bar }

SerializableEnumeration(MyEnum);

// in my.model.ts
export class MyModel {

    @deserializeAs(MyEnum)
    typeEnum: MyEnum;
}
```

`MyEnum` is supposed to be an `ISerializable`. With `SerializableEnumeration()`, it is, at runtime. But at compilation type, Typescript has no idea about this extension and the new weak type checking throws.

Typescript proposes a workaround at the end of the section.

> Since this is a breaking change, you may need to know about the workarounds which are the same as those for strict object literal checks:
> 
> 1. Declare the properties if they really do exist.
> 2. Add an index signature to the weak type (i.e. [propName: string]: {}).
> 3. Use a type assertion (i.e. opts as Options).

**Comment by ThePainnn** [comment]
HI, I'm new to typescript and all. I've got this issue now with my enum annotated with serializeauto. Is there a workaround or we have to wait for an updated version of the lib for that issue?

Thanks!

**Comment by noemi-salaun** [comment]
@ThePainnn 
You can go back to TypeScript 2.3 until @weichx looks at the problem.

**Comment by weichx** [comment]
I'll have to think about how to handle this, I'm not sure off the top of my head

**Comment by weichx** [comment]
I think the only solution is have SerializableEnum return an indexable mapped type. I forget exactly how mapped types work but it would something like this:

````
function SerializableEnum<T extends Indexable<number|string>>(_enum : T) : {[K in keyof T] : number|string}

const EnumType = SerializableEnum<MyEnum>(enum Thing { ... });

export class MyModel {
    @deserializeAs(EnumType) typeEnum: EnumType;
}

````

Thoughts?

**Comment by rsparrow** [comment]
+1 Look forward to seeing how this is resolved ;-)

**Comment by 7chenko** [comment]
Any resolution on this? Thanks.

**Comment by weichx** [comment]
Should be fixed by #71
